src/host/getset.cpp

```diff
@@ -1417,18 +1417,45 @@ void DoSrvPrivateAllowCursorBlinking(SCREEN_INFORMATION& screenInfo, const bool
     // Make sure the cursor doesn't move outside the viewport.
     screenInfo.GetViewport().Clamp(clampedPos);
 
-    // Make sure the cursor stays inside the margins, but only if it started there
-    if (screenInfo.AreMarginsSet() && screenInfo.IsCursorInMargins(cursor.GetPosition()))
+    // Make sure the cursor stays inside the margins
+    if (screenInfo.AreMarginsSet())
     {
-        try
+        const auto margins = screenInfo.GetAbsoluteScrollMargins().ToInclusive();
+
+        const auto newY = clampedPos.Y;
+        const auto cursorY = cursor.GetPosition().Y;
+
+        const auto lo = margins.Top;
+        const auto hi = margins.Bottom;
+
+        // See microsoft/terminal#2929 - If the cursor is _below_ the top
+        // margin, it should stay below the top margin. If it's _above_ the
+        // bottom, it should stay above the bottom. Cursor movements that stay
+        // outside the margins shouldn't necessarily be affected. For example,
+        // moving up while below the bottom margin shouldn't just jump straight
+        // to the bottom margin. See
+        // ScreenBufferTests::CursorUpDownOutsideMargins for a test of that
+        // behavior.
+        const bool cursorBelowTop = cursorY > lo;
+        const bool cursorAboveBottom = cursorY < hi;

```

Comment:
I think these are both off by one. If the cursor Y position exactly matches the top margin, it is still "inside" that margin, and shouldn't be able to move up. Similarly, if it exactly matches the bottom margin, it shouldn't be able to move down.